muxink/src/demux.rs

```diff
@@ -0,0 +1,325 @@
+        Demultiplexer {
+            // We fuse the stream in case its unsafe to call it after yielding `Poll::Ready(None)`
+            stream: stream.fuse(),
+            // Initially, we have no next frame
+            next_frame: None,
+            // Initially, all channels are inactive
+            active_channels: [0b00000000; 32],
+        }
+    }
+}
+
+// Here, we write the logic for accessing and modifying the bit-field representing the active
+// channels.
+impl<S> Demultiplexer<S> {
+    fn activate_channel(&mut self, channel: u8) {
+        self.active_channels[(channel / 8) as usize] |=
+            2u8.checked_pow((channel % 8) as u32).unwrap();
+    }
+
+    fn deactivate_channel(&mut self, channel: u8) {
+        // TODO Single operation instead of two.
+        if self.channel_is_active(channel) {
+            self.active_channels[(channel / 8) as usize] ^=
+                2u8.checked_pow((channel % 8) as u32).unwrap();
+        }
+    }
+
+    fn channel_is_active(&self, channel: u8) -> bool {
+        (self.active_channels[(channel / 8) as usize]
+            & 2u8.checked_pow((channel % 8) as u32).unwrap())
+        .count_ones()
+            == 1
+    }
+}
+
+impl<S: Stream> Demultiplexer<S> {
+    /// Creates a handle listening for frames on the given channel.
+    ///
+    /// Any item on this channel sent to the `Stream` underlying the `Demultiplexer` we used to
+    /// create this handle will be read only when all other messages for other channels have been
+    /// read first. If one has handles on the same channel created via the same underlying
+    /// `Demultiplexer`, each message on that channel will only be received by one of the handles.
+    /// Unless this is desired behavior, this should be avoided.
+    pub fn create_handle(demux: Arc<Mutex<Self>>, channel: u8) -> DemultiplexerHandle<S> {
+        let mut guard = match demux.as_ref().try_lock() {
+            Err(_err) => panic!("TODO"),
+            Ok(guard) => guard,
+        };
+
+        if guard.channel_is_active(channel) {
+            panic!("TODO")
+        }
+
+        guard.activate_channel(channel);
+
+        DemultiplexerHandle {
+            channel,
+            demux: demux.clone(),
+        }
+    }
+}
+
+/// A handle to a demultiplexer.
+///
+/// A handle is bound to a specific channel, see [`Demultiplexer::create_handle`] for details.
+pub struct DemultiplexerHandle<S> {
+    /// Which channel this handle is listening on
+    channel: u8,
+    /// A reference to the underlying demultiplexer.
+    demux: Arc<Mutex<Demultiplexer<S>>>,
+}
+
+impl<S> Drop for DemultiplexerHandle<S> {
+    fn drop(&mut self) {
+        let mut demux = match self.demux.as_ref().try_lock() {
+            Err(_err) => {
+                return;
+            } // TODO What do? Perhaps try_lock is wrong here, but still what about poisoning? Not doing anything seems like the
+            // only sane option
+            Ok(guard) => guard,
+        };
```

Comment:
This goes away once you separate locking from the rest of the logic. This would be a blanket `Drop` impl from the compiler for the `Arc<Mutex>>` part, and then the code below would be the `Drop` impl for the self-contained object, if that makes sense.

Reply:
I disagree. If you hold a `DemultiplexerHandle`, you will need a reference to the "original" demultiplexer in some shape or form. Said reference must be protected through a lock.

The code here is responsible for deregistering the channel from said demultiplexer. If another thread has poisoned the mutex, we have an error and there is nothing we can to.

`try_lock`, however, *is* wrong here, as it just makes the locking non-blocking (coincidentally, I fell into that same trap a few months ago...). We *do* want to wait to deregister in normal operation.

Reply:
I left this comment a while back, I don't remember exactly what I was thinking about back then but I think it would have been something like move the decrementing of the ref counter to the higher level object and not have it be in a `Drop` impl at all, have it done manually so you can decide what to do in case the lock is poisoned when you try to free the resource. This was all to avoid the locking in `drop`, but after reading your comment, I'm not sure it would add that much value so let's go with the original.

> try_lock, however, is wrong here, as it just makes the locking non-blocking (coincidentally, I fell into that same trap a few months ago...). We do want to wait to deregister in normal operation.

Yes, I missed it as well :sweat_smile:. I agree.